### PR TITLE
Add get_conv_idx for convolution layer roasting

### PR DESCRIPTION
**Pull request by kimiasa**

### CODE REVIEW

File: Mapper_v2.py [hunk 1]
```diff
@@ -212,6 +220,11 @@ def get_mlp_idx(self, w_shape, original_offset, target_size, block_k, block_n, *
       #print("RoastMemOptMapper get_mlp_idx")
       #print(idx[:5,:5])
       return idx
+    
+    def get_conv2d_idx(self, w_shape, original_offset, target_size, block_k, block_n, **kwargs):
+        assert(len(w_shape) == 4)
+        idx = self.get_mlp_idx([w_shape[0], np.prod(w_shape[1:])], original_offset, target_size, block_k, block_n, **kwargs)
```
Comment (apd10) [inline Mapper_v2.py:230]: I think according to the implicit gemm dimensions here : https://docs.nvidia.com/deeplearning/performance/dl-performance-convolutional/index.html for forward convolution, we should do get mlp on 
(wshape[0]*wshape[2]*wshape[3] , wshape[1])

File: Mapper_v2.py [hunk 2]
```diff
@@ -240,6 +253,11 @@ def get_mlp_idx(self, w_shape, original_offset, target_size, block_k, block_n, b
       #print("RoastCompOptMapper get_mlp_idx")
       #print(idx[:5,:5])
       return idx
+    
+    def get_conv2d_idx(self, w_shape, original_offset, target_size, block_k, block_n, block_k_small, **kwargs):
+        assert(len(w_shape) == 4)
+        idx = self.get_mlp_idx([w_shape[0], np.prod(w_shape[1:])], original_offset, target_size, block_k, block_n, block_k_small, **kwargs)
+        return idx.view(*w_shape)
```
Comment (apd10) [inline Mapper_v2.py:265]: same as above

**Review by apd10 — CHANGES_REQUESTED** [review]
Added some comments. I don't think it is the correct layout.

File: Mapper_v2.py (hunk 1 above)
Comment (kimiasa) [reply Mapper_v2.py:230]: This is the weight tensor `W_shape = (out_channels, int(in_channels/groups),
                   kernel_size[0], kernel_size[1])` so the flattend dimensions would be correct. You are right about the dimensions I should swap dims as follow: `(wshape[1]*wshape[2]*wshape[3] , wshape[0])`

Comment (kimiasa) [reply Mapper_v2.py:230]: However I think I should not swap actually because in mlp matrix mult the weight tensor is `self.W_shape = (output, input)`

Comment (apd10) [reply Mapper_v2.py:230]: Even then it should be (wshape[1],  wshape[0]*wshape[2]*wshape[3])

Comment (kimiasa) [reply Mapper_v2.py:230]: I am not convinced. I read the RzLinear the reason we write the output channel as zeroth dimension of the matrix is that in memory the Cuda will write weights column majored. The K is the number of columns. Based on that, we should use block_k for chunking input _channels or `w_shape[1]` and block_n  for chuncking output_channels or `w_shape[0]`. I referred to the following lines of code in RzLinear: `float * weights, //  K x N (column major)
    float * input, // M x K (row major)
    float * output, // M x N (row major)`

**Comment by apd10** [comment]
Lets discuss tomorrow in person if you are around.

**Comment by kimiasa** [comment]
Sounds good! I appreciate if you send meeting details tomorrow. Currently, I am watching the Nvidia implicit GEMM implementation to learn the process, might prepare some slides.

**Pull request merged**